lib/ts/recipe/multifactorauth/types.ts

```diff
@@ -0,0 +1,129 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { BaseRequest, BaseResponse } from "../../framework";
+import OverrideableBuilder from "supertokens-js-override";
+import { GeneralErrorResponse } from "../../types";
+import { SessionContainer } from "../session";
+import { SessionContainerInterface } from "../session/types";
+
+export type MFARequirement =
+    | {
+          id: string;
+          maxAgeInSeconds?: number;
+      }
+    | string;
+
+export type MFARequirementList = (
+    | {
+          oneOf: MFARequirement[];
+      }
+    | {
+          allOf: MFARequirement[];
+      }
+    | MFARequirement
+)[];
+
+export type MFAClaimValue = {
+    c: Record<string, number>;
+    n: string[];
+};
+
+export type TypeInput = {
+    firstFactors?: string[];
+
+    getMFARequirementsForAuth?: (
+        session: SessionContainer | undefined,
+        factorsSetUpByTheUser: string[],
+        completedFactors: Record<string, number>,
+        userContext: any
+    ) => Promise<MFARequirementList> | MFARequirementList;
+    getMFARequirementsForFactorSetup?: (
+        factorId: string,
+        session: SessionContainer,
+        factorsSetUpByTheUser: string[],
+        completedFactors: Record<string, number>,
+        userContext: any
+    ) => Promise<MFARequirementList> | MFARequirementList;
+
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    firstFactors?: string[];
+
+    getMFARequirementsForAuth: (
+        session: SessionContainer | undefined,
+        factorsSetUpByTheUser: string[],
+        completedFactors: Record<string, number>,
+        userContext: any
+    ) => Promise<MFARequirementList> | MFARequirementList;
+    getMFARequirementsForFactorSetup: (
+        factorId: string,
+        session: SessionContainer,
+        factorsSetUpByTheUser: string[],
+        completedFactors: Record<string, number>,
+        userContext: any
+    ) => Promise<MFARequirementList> | MFARequirementList;
+
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type RecipeInterface = {
+    isAllowedToSetupFactor: (input: {
+        session: SessionContainerInterface;
+        factorId: string;
+        userContext: any;
+    }) => Promise<boolean>;
+    getFactorsSetupForUser: (input: { userId: string; tenantId: string; userContext: any }) => Promise<string[]>;
```

Comment:
```suggestion
    getFactorsSetupForUser: (input: { session: SessionContainer; userContext: any }) => Promise<string[]>;
```

Reply:
Isn't there a use-case where a dev wants to call this to get info about an arbitrary user? We may event want to call it like that our dashboard impl.

Reply:
I think that's a good compromise.

Reply:
I'll make this change since I think it's fine, but note that we are breaking the pattern of userrules and EV where we accept any user id